File: Algorithm-Swift/Algorithm-Swift/Inflearn/54_올바른괄호.swift
//
//  54_올바른괄호.swift
//  Algorithm-Swift
//
//  Created by SHIN YOON AH on 2023/08/26.
//

import Foundation

func solution54() {
    let input = readLine()!
    var stack: [String] = []

    for n in input.map({ String($0) }) {
        if n == "(" { stack.append("(") }
        else {
            if stack.isEmpty {
                print("NO")
                return
            } else {
                stack.removeLast()
            }
        }
    }

    if stack.isEmpty { print("YES") }
    else { print("NO") }
}

File: Algorithm-CPP/54_올바른괄호.cpp
#include <iostream>
#include <stack>
using namespace std;

int main() {
    string input;
    stack<char> stack;

    cin >> input;

    for(int i = 0; i < input.size(); i++) {
        if(input[i] == '(') stack.push('(');
        else {
            if(!stack.empty()) stack.pop();
            else {
                cout << "NO" << endl;
                return 0;
            }
        }
    }

    if(stack.empty()) cout << "YES" << endl;
    else cout << "NO" << endl;

    return 0;
}